polars/polars-io/src/ipc.rs

```diff
@@ -274,6 +273,115 @@ where
     }
 }
 
+#[must_use]
+#[cfg(feature = "partition")]
+/// Write a DataFrame to partitioned Arrow's IPC format
+/// # Example
+///
+/// ```
+/// use polars_core::prelude::*;
+/// use polars_core::df;
+/// use polars_io::ipc::PartitionIpcWriter;
+///
+/// fn example(df: &DataFrame) -> Result<()> {
+///     let df = df!("a" => [1, 1, 2, 3], "b" => [2, 2, 3, 4], "c" => [2, 3, 4, 5]).unwrap();
+///     let by = ["a", "b"];
+///     PartitionIpcWriter::new("./tmp_dir", by)
+///         .finish(&df)
+/// }
+/// ```
+pub struct PartitionIpcWriter<P, I, S> {
+    rootdir: P,
+    by: I,
+    parallel: bool,
+    compression: Option<write::Compression>,
+    phantom: PhantomData<S>,
+}
+
+#[cfg(feature = "partition")]
+impl<P, I, S> PartitionIpcWriter<P, I, S>
+where
+    P: Into<PathBuf>,
+    I: IntoIterator<Item = S>,
+    S: AsRef<str>,
+{
+    pub fn new(rootdir: P, by: I) -> Self {
+        Self {
+            rootdir,
+            by,
+            parallel: true,
+            compression: None,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Set the compression used. Defaults to None.
+    pub fn with_compression(mut self, compression: Option<write::Compression>) -> Self {
+        self.compression = compression;
+        self
+    }
+
+    /// Write the ipc file in parallel (default). The single threaded writer consumes less memory.
+    pub fn write_parallel(mut self, parallel: bool) -> Self {
+        self.parallel = parallel;
+        self
+    }
+
+    pub fn finish(self, df: &DataFrame) -> Result<()> {
+        use polars_core::POOL;
+        use rayon::iter::IntoParallelIterator;
+        use rayon::iter::ParallelIterator;
+        use uuid::Uuid;
+
+        let rootdir: PathBuf = self.rootdir.into();
+        let by: Vec<String> = self.by.into_iter().map(|x| x.as_ref().to_owned()).collect();
+        let group = df.groupby(&by)?;
```

Comment:
We can use existing `partition_by` operation here.

```rust
        let partitioned = df.partition_by(by)?;
        partitioned.iter().for_each(|df| {
            path.push(format!("part-{}.ipc", Uuid::new_v4()));

            let mut file = std::fs::File::create(path)?;
            let writer = BufWriter::new(&mut file);

            IpcWriter::new(writer)
                .with_compression(self.compression)
                .finish(&mut partition_df)
        });

```

Reply:
I know `partitioned_by` function, but this function collects all of the DataFrames. I think this logic does not collect all, so reducing memory allocation.

Reply:
Ok, I think we should then adapt `partition_by` a little bit by letting that return `Result<impl Iterator<Item=DataFrame>>`. That way we can reduce the amount of code and keep it more DRY.

Reply:
Cool. It seems the best way.

Reply:
I used `partition_by` this time